src/core/consent/createConsentState.js

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+import { cookieJar, getNamespacedCookieName, values } from "../../utils";
+import parseConsentCookie from "./parseConsentCookie";
+import { CONSENT_COOKIE_KEY } from "../../constants/cookieDetails";
+import { IN } from "../../constants/consentStatus";
+
+const hasConsentedToAllPurposes = consentByPurpose => {
+  return (
+    consentByPurpose && values(consentByPurpose).every(value => value === IN)
+  );
+};
+
+const getPersistedConsent = cookieName => {
+  const cookieValue = cookieJar.get(cookieName);
+  if (cookieValue) {
+    return parseConsentCookie(cookieValue);
+  }
+  return undefined;
+};
+
+export default ({ config }) => {
+  const { orgId } = config;
+  const consentCookieName = getNamespacedCookieName(orgId, CONSENT_COOKIE_KEY);
+  const onChangeHandlers = [];
+  let consentByPurpose;
+  let suspended = false;
+
+  const setConsentByPurpose = _consentByPurpose => {
+    consentByPurpose = _consentByPurpose;
+    onChangeHandlers.forEach(onChangeHandler => {
+      onChangeHandler();
+    });
+  };
+  const updateFromCookies = () => {
+    if (!suspended) {
+      setConsentByPurpose(getPersistedConsent(consentCookieName));
+    }
+  };
+
+  updateFromCookies();
+
+  return {
+    suspend() {
+      if (!suspended) {
+        suspended = true;
+        setConsentByPurpose(undefined);
+      }
+    },
+    unsuspend() {
+      if (suspended) {
+        suspended = false;
+        updateFromCookies();
+      }
+    },
+    updateFromCookies,
+    isPending() {
+      return consentByPurpose === undefined;
+    },
+    // TODO Once we support consenting to specific purposes, this
+    // method will accept an array of purpose names as an argument and
+    // will return whether the user has consented into those purposes.
+    hasConsentedToAllPurposes() {
+      return (
+        Boolean(consentByPurpose) && hasConsentedToAllPurposes(consentByPurpose)
+      );
+    },
+    onChange(onChangeHandler) {
```

Comment:
I'm not a big fan of the onChange event here.  I think it makes things unnecessarily complicated, especially when awaitConsent is the only thing that uses it.  Do you think this would be used by other things?

The only time awaitConsent needs to possibly resolve promises is after `unsuspend` or `updateFromCookies` is called on the Consent object.  So the consentObject could notify awaitConsent when this happens. (awaitConsentFactory would have to return an object instead of a function so that it could expose processConsent, and awaitConsent)

Reply:
I have conflicting opinions on this. I believe `awaitConsent` should always process the consent state whenever it changes and the most robust way to do this is to listen directly to the state itself instead of being told be intermediaries. By having `awaitConsent` rely on any modules that are _not_ the state to tell it about potential changes to state, we're relying on future developers to understand that relationship and remember to notify `awaitConsent` whenever state is updated.

For example, if we make the consent object tell awaitConsent when to process consent, but then we add something else in the system that calls `consentState.unsuspend` or `consentState.updateFromCookies` but forget to tell awaitConsent about it, then consent breaks. It seems proper to have `awaitConsent` be notified from the source of record, which is the consent state itself.

Another option would be to send awaitConsent into consentState instead of the other way around. Then consentState would a method directly on awaitConsent. I feel like it's not really in the state's domain to know which other modules care about when the state changes, but I'd be open to it.

Reply:
I didn't answer your question though: 

> Do you think this would be used by other things?

Probably not.

Reply:
👍